[CARBON-865] CustomDragLayer component for enabling Drag & Drop ghost row

paulsturgess opened the pull request:
Ticket CARBON-865

Implemented by using: https://react-dnd.github.io/react-dnd/docs-drag-layer.html 

**Chrome**
![drag-chrome](https://user-images.githubusercontent.com/4964/29129945-565bc002-7d20-11e7-95d7-ef5cd44f034f.gif)

**Firefox**
![drag-ff](https://user-images.githubusercontent.com/4964/29130056-a41c68f0-7d20-11e7-990a-12efed6b0da9.gif)

**Edge**
![drag-edge](https://user-images.githubusercontent.com/4964/29130443-c574e710-7d21-11e7-859f-ec5a7afb45b3.gif)


TODO:
- [x] Figure out how we can handle the width of the ghost row
- [x] Specs
- [x] Add the ghost row to the drag and drop demo
- [x] Changelog

SeanArmstrong commented:
Went to have a play with it and got this error - https://github.com/react-dnd/react-dnd/issues/431

Caused by this - https://github.com/react-dnd/react-dnd/blob/ca10e688651833d10b378a2a1bf284a0264ed4c8/packages/react-dnd/src/DragLayer.js#L96-L105 

Possible workaround at the bottom of the page

paulsturgess commented:
@SeanArmstrong thanks for the heads up, I've managed to remove the error. I've updated the `CustomDragLayer` to pass through a prop of `customDragLayer: true`. This allows the developer to change the behaviour of the component used for the ghost row (the child of `CustomDragLayer`).

So for the `ConfigurableItemRow`, I have updated it to check if the prop is true, and if so, remove any references to `WithDrop` and `WithDrag`.

It might still be worth splitting `ConfigurableItemRow` into two components, one draggable and one not draggable so that it is not littered with `if (this.props.customDragLayer)`. I'd need to split the icon and the list item into stand-alone components, to make them easy to compose.

paulsturgess commented:
Work on this has paused due to the difficulty of placing the ghost layer inside the table which renders the html markup invalid.

paulsturgess commented:
This has now been refactored so that the `<CustomDragLayer />` can be defined outside the table, which means the markup remains valid.

Tested on the following browsers:
- Mac Chrome, Safari & Firefox
- Windows 8.1 IE 11
- Windows 10 MSEdge

paulsturgess commented:
Note that when dragging on some elements the mouse cursor changes (e.g. hover over text and it no longer shows the grabbing hand cursor). I've spent some time trying to resolve this but it appears it's not possible due to browser limitations. Confirmed in this issue on the repo: https://github.com/react-dnd/react-dnd/issues/325

paulsturgess commented:
Unfortunately Safari is not displaying the 'grabbing' cursor when dragging and instead shows the text cursor.

paulsturgess commented:
I have raised separate tickets for the issue with the Safari cursor and the performance of the DraggableContext component on the Carbon site. CARBON-942 and CARBON-939.

SeanArmstrong commented on src/components/drag-and-drop/draggable-context/draggable-context.js:
```diff
@@ -146,7 +148,12 @@ class DraggableContext extends React.Component {
+    return (
+      <div className='carbon-draggable-context'>
+        { this.props.children }
+        <CustomDragLayer />
```
Could this pass some DragLayerProps so that its easier to change the custom drag layer.

Or enable a way so that you can optionally pass your own custom drag layer.

paulsturgess replied:
Yeah I like the idea of being able to pass in your own drag layer into the Draggable Context.

paulsturgess commented:
@SeanArmstrong I've updated the `DraggableContext` component so the CustomDragLayer can be optionally passed in as a prop.

SeanArmstrong commented on src/components/drag-and-drop/draggable-context/draggable-context.js:
```diff
@@ -142,11 +152,23 @@ class DraggableContext extends React.Component {
+    if (this.props.customDragLayer) {
+      return this.props.customDragLayer;
+    }
+    return <CustomDragLayer />;
```
As this is not adding any additional props could it just be defined as a default prop removing the need for this method. Not sure exactly how this works for components

paulsturgess commented:
@SeanArmstrong `<CustomDragLayer />` is now a default prop.

ianrobsonsage commented:
QA'd this in the following browsers:
Chrome 
Safari 
Firefox 
Edge 
All working as expected, other than the issue already raised as CARBON-939

andrewjtait commented:
![screen shot 2017-08-18 at 11 21 25](https://user-images.githubusercontent.com/1668080/29455020-6134f4b0-8407-11e7-8be5-063ce1e45481.png)

Can the row below the dragged row get a border top?

andrewjtait commented on src/components/drag-and-drop/custom-drag-layer/custom-drag-layer.js:
```diff
@@ -0,0 +1,121 @@
+      <div className={ this.classes }>
+        <div
+          className='custom-drag-layer__container'
+          ref={ this.setContainerRef }
```
can just have this as an inline function for consistency?

andrewjtait commented on src/components/drag-and-drop/custom-drag-layer/custom-drag-layer.js:
```diff
@@ -0,0 +1,121 @@
+     * @property className
+     * @type {String}
+     */
+    className: PropTypes.string,
```
does this ever receive `className` as a prop?

paulsturgess replied:
It doesn't i'll remove.

andrewjtait commented on src/components/drag-and-drop/draggable-context/__definition__.js:
```diff
@@ -23,18 +22,21 @@ let definition = new Definition('draggable-context', DraggableContext, {
-  </DraggableContext>`
+    onDrag: 'updateDndData',
+    children: `
+    <div>
```
indentation is wrong on here:
![screen shot 2017-08-18 at 11 37 26](https://user-images.githubusercontent.com/1668080/29455587-9ad420f4-8409-11e7-88e2-d63e99dac5ec.png)

andrewjtait commented on src/components/drag-and-drop/custom-drag-layer/custom-drag-layer.js:
```diff
@@ -0,0 +1,121 @@
+  return {
+    currentOffset: monitor.getSourceClientOffset(),
+    item,
+    draggableNode: item ? item.draggableNode : null,
```
should this be listed as a prop type in `WithDrag` so it's clear where you set it?

andrewjtait commented on src/components/drag-and-drop/custom-drag-layer/custom-drag-layer.js:
```diff
@@ -0,0 +1,121 @@
+
+  componentDidUpdate(prevProps) {
+    if (this.props.draggableNode && (prevProps.draggableNode !== this.props.draggableNode)) {
+      this.setClonedChildWidth();
```
should we only do this when we need to (when we're dragging or creating the clone?)

andrewjtait commented on src/components/drag-and-drop/custom-drag-layer/custom-drag-layer.js:
```diff
@@ -0,0 +1,121 @@
+    }
+    if (this.props.draggableNode && this.props.isDragging) {
+      this.createClonedChild();
+    }
```
would we ever be in the scenario of needing this on mount?

paulsturgess replied:
I've removed this 👍

paulsturgess commented:
@andrewjtait Feedback all addressed. Thanks!

andrewjtait commented on src/components/configurable-items/configurable-item-row/configurable-item-row.js:
```diff
@@ -62,6 +62,10 @@ class ConfigurableItemRow extends React.Component {
 
+  state = {
+    draggableNode: null
+  }
```
this can be removed

ianrobsonsage commented:
Re-QA'd after further improvements have been made.

As previously fine in all browsers other than the Safari icon issue already logged. QA Passed.
